File: leetcode/linked-list/206_reverse_linked_list.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

// 解题思路：使用头插法，需要注意的是 golang 中的 怎么声明 nil 的指针
// 时间复杂度 O(n)

func reverseList(head *ListNode) *ListNode {
	var dummyP *ListNode
	tmpP := &ListNode{}
	for head != nil {
		tmpP = head.Next
		head.Next = dummyP
		dummyP = head
		head = tmpP
	}
	return dummyP
}

// Follow up:
// A linked list can be reversed either iteratively or recursively. Could you implement both?
// 既然提示了可以用迭代和递归解，再来试试递归

File: leetcode/linked-list/206_reverse_linked_list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        rev_head = None
        while head is not None:
            tmp = head.next
            head.next = rev_head
            rev_head = head
            head = tmp
        return rev_head
     
    def reverseList(self, head, last = None):
    	if not head:
            return last
    	next = head.next
    	head.next = last
    	return self.reverseList(next, head)
